ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,332 @@
+
+  private final KsqlEngine engine;
+  private final ServiceContext serviceContext;
+  private final ConfigurationParameters configs;
+  private final ScheduledExecutorService scheduledExecutorService;
+  private final ConcurrentHashMap<HostInfo, TreeMap<Long, HeartbeatInfo>> receivedHeartbeats;
+  private final ConcurrentHashMap<String, Boolean> aliveHosts;
+  private final ConcurrentHashMap<URI, Boolean> remoteHostsURI;
+  private final ConcurrentHashMap<HostInfo, Boolean> remoteHostsInfo;
+  private final HostInfo localHostInfo;
+  private final URL localURL;
+
+
+  public static HeartbeatHandler.Builder builder() {
+    return new HeartbeatHandler.Builder();
+  }
+
+  private HeartbeatHandler(KsqlEngine engine,
+                           ServiceContext serviceContext,
+                           URI localServerURI,
+                           ConfigurationParameters configs) {
+
+    this.engine = requireNonNull(engine, "engine");
+    this.serviceContext = requireNonNull(serviceContext, "serviceContext");
+    this.configs = requireNonNull(configs, "configuration parameters");
+    this.scheduledExecutorService = Executors.newScheduledThreadPool(configs.threadPoolSize);
+    this.receivedHeartbeats = new ConcurrentHashMap<>();
+    this.aliveHosts = new ConcurrentHashMap<>();
+    this.remoteHostsURI = new ConcurrentHashMap<>();
+    this.remoteHostsInfo = new ConcurrentHashMap<>();
+    this.localHostInfo = new HostInfo(localServerURI.getHost(), localServerURI.getPort());
+    try {
+      this.localURL = localServerURI.toURL();
+    } catch (final Exception e) {
+      throw new IllegalStateException("Failed to convert local host URI to URL."
+                                          + " local host URL: " + localServerURI);
+    }
+    this.aliveHosts.putIfAbsent(localHostInfo.toString(), true);
+  }
+
+  void scheduleHeartbeatRelatedTasks() {
+    SendHeartbeatTask sendTask = new SendHeartbeatTask();
+    scheduledExecutorService.scheduleAtFixedRate(sendTask, 0, configs.heartbeatSendIntervalMs,
+                                                 TimeUnit.MILLISECONDS);
+    CheckHeartbeatTask processTask = new CheckHeartbeatTask();
+    scheduledExecutorService.scheduleAtFixedRate(processTask, 0, configs.heartbeatCheckIntervalMs,
+                                                 TimeUnit.MILLISECONDS);
+    DiscoverClusterTask discoverTask = new DiscoverClusterTask();
+    scheduledExecutorService.scheduleAtFixedRate(discoverTask, 0, configs.discoverClusterIntervalMs,
+                                                 TimeUnit.MILLISECONDS);
+
+  }
+
+  void shutDown() {
+    scheduledExecutorService.shutdown();
+  }
+
+  public ConcurrentHashMap<HostInfo, TreeMap<Long, HeartbeatInfo>> getReceivedHeartbeats() {
+    return receivedHeartbeats;
+  }
+
+  ConcurrentHashMap<URI, Boolean> getRemoteHostsURI() {
+    return remoteHostsURI;
+  }
+
+  public Map<String, Boolean> getAliveHosts() {
+    return aliveHosts;
```

Comment:
for all of these you probably want to expose an Immutable copy or view of whatever the internal state is - others should not be modifying it